File: src/ocaml/src/sniper.ml
#load "str.cma"

#use "./src/ocaml/utils/tester.ml"
#use "./src/ocaml/utils/print_util.ml"
#use "./src/ocaml/utils/about_util.ml"

module S = Str module L = List module St = String

let joining = "sniper item-123 1300 0 0 JOINING"
let bidding = "sniper item-123 1300 1000 1098 BIDDING"
let winning = "sniper item-123 1300 1098 1098 WINNING"
let close_msg = "SOLVersion: 1.1; Event: CLOSE;"
let price_msg = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 1098; Increment: 24; Bidder: other bidder;"
let price_me = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 1098; Increment: 24; Bidder: sniper;"
let price_over = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 1280; Increment: 24; Bidder: some one;"
let price_not_val = "SOLVersion: 1.1; Event: PRICE; CurrentPrice:; Increment: 24; Bidder: other bidder;"
let price_not_CP = "SOLVersion: 1.1; Event: PRICE; Increment: 24; Bidder: other bidder;"
let price_not_EV = "SOLVersion: 1.1; CurrentPrice: 1098; Increment: 24; Bidder: other bidder;"
let no_msg = "toAuctionMessage is None"
let bid_msg = "SOLVersion: 1.1; Command: BID; Price: "

let lines str = let ss = split str "\n" in (L.nth ss 0, L.nth ss 1)

let words str = let ss = split str " " in (L.nth ss 0, L.nth ss 1, L.nth ss 2, L.nth ss 3, L.nth ss 4, L.nth ss 5)

let ret_to_sp ss = let (a,b) = lines ss in (St.trim a ) ^ " " ^ (St.trim b)

let test_lw() = let (a,b) = lines (joining ^ "\n" ^ close_msg) and (c,d,e,f,g,h) = words joining
  in ps (a^b^c^d^e^f^g^h)

type sniper = 
  {name : string; item : string; stop_price : int; last_price : int; last_bid : int; state : string; auc_msg : string}

let str_sniper {item = i; last_price = lp; last_bid = lb; state = st; auc_msg = am} = 
  am ^ "\n" ^ i ^ " " ^ to_s lp ^ " " ^ to_s lb ^ " " ^ st 

let mk_sniper (a,b,c,d,e,f) =
  {name = a; item = b; stop_price = to_i c; last_price = to_i d; last_bid = to_i e; state = f; auc_msg = no_msg}

let test_s() = let j = mk_sniper(words joining) in ps (str_sniper j)


type translator = Price of int * int * string | Failed of string | Close

let str_trans = function Price (a,b,c) -> to_s a ^ " " ^ to_s b ^ " " ^ c | Failed s -> "failed: " ^ s | close -> "close"

let test_st() = let res = str_trans (Price (25, 43, "bidder")) in ps res

let lack = Failure "not enough args" and empty s = Failure ("empty field val is " ^ s)
  and missing s = Failure ("missing field is " ^ s)

let semi_colons str = split str ";" and colons str = split str ":"

let safe_colons = function ([] | _::[]) -> raise lack | a::b::_ -> (St.trim a, St.trim b)

let pairs msg =
  let rec inner = function [] -> [] | x::xs -> safe_colons(colons x) :: inner xs in inner (semi_colons msg)

let test_ps() = let res = pairs price_msg in L.iter(fun (a,b) -> ps (a ^ " " ^ b)) res

let is_empty k v = if v = "" then raise (empty k) else v

let rec find_msg key = function [] -> raise (missing key)
  | (k,v) :: xs -> if k = key then is_empty k v else find_msg key xs

let event msg = let msgs = pairs msg in let ev = find_msg "Event" msgs in match ev with
  | "CLOSE" -> Close
  | "PRICE" -> Price (to_i(find_msg "CurrentPrice" msgs), to_i(find_msg "Increment" msgs), (find_msg "Bidder" msgs))
  | _       -> Failed "Event field is notFound"

let test_ev() = let res = event close_msg in ps (str_trans res)

let translate msg = try event msg with Failure e -> Failed ("<" ^ msg ^ ">  [" ^ e ^ "]")

let test_t() = let res = translate price_not_EV in ps (str_trans res)


let close s = if s.state = "WINNING" then {s with state = "WON"} else {s with state = "LOST"}

let price s p i b = if b = s.name then {s with last_price = p; state = "WINNING"} else
  let bid = p + i in if bid > s.stop_price then {s with last_price = p; state = "LOSING"} else
  {s with last_price = p; last_bid = bid; state = "BIDDING"; auc_msg = bid_msg ^ to_s bid ^ ";"}

let failed s m = {s with last_price = 0; last_bid = 0; state = "FAILED"; auc_msg = "failed: " ^ m}

let job s = function Close -> close s | Price (a,b,c) -> price s a b c | Failed m -> failed s m

let domain (items, msg) =
  let s = mk_sniper (words items) and t = translate msg in let res = job s t in str_sniper res

let test() = let res = domain (bidding, price_me) in ps res

let refactor() = let arg = read_to_file "a:/pj/sniper/rr/req1.txt" and exp = read_to_file "a:/pj/sniper/rr/res2.txt" in
  let res = domain (lines arg) in
    zz_act ~dsp: "WhenJoinningIfReceivedCloseInfoWhileNoMsgAndLost by Refactor" "Sniper Test" (ret_to_sp res) (ret_to_sp exp)

let developer arg exp name = zz_act ~dsp:name "Sniper Test" (ret_to_sp(domain(lines arg))) (ret_to_sp exp)
let develop() =
  let req1 = read_to_file "a:/pj/sniper/rr/req1.txt" and res1 = read_to_file "a:/pj/sniper/rr/res1.txt" in
  let req2 = read_to_file "a:/pj/sniper/rr/req2.txt" and res2 = read_to_file "a:/pj/sniper/rr/res2.txt" in
  let req3 = read_to_file "a:/pj/sniper/rr/req3.txt" and res3 = read_to_file "a:/pj/sniper/rr/res3.txt" in
  let req4 = read_to_file "a:/pj/sniper/rr/req4.txt" and res4 = read_to_file "a:/pj/sniper/rr/res4.txt" in
  let req5 = read_to_file "a:/pj/sniper/rr/req5.txt" and res5 = read_to_file "a:/pj/sniper/rr/res5.txt" in
  let req6 = read_to_file "a:/pj/sniper/rr/req6.txt" and res6 = read_to_file "a:/pj/sniper/rr/res10.txt" in
  developer req1 res1 "WhenReceivedBuiableInfoThenBuyAndJoining";
  developer req2 res2 "WhenJoinningIfReceivedCloseInfoWhileNoMsgAndLost";
  developer req3 res3 "WhenBiddeingIfReceivedCloseInfoWhileMoMsgAndLost";
  developer req4 res4 "WhenWinningIfReceivedCloseInfoWhileNoMsgAndWon";
  developer req5 res5 "WhenWinningIfReceivedTooHiPriceInfoWhileNoMsgAndLosing";
  developer req6 res6 "WhenWinningIfReceivedFunnyMsgWhileFailMsgAndFailed"

let product() = let arg = read_to_file "toSniper.txt" in write_to_file "fromSniper.txt" (domain(lines arg))


let () = 
  (* test() *)
  (* refactor() *)
  (* develop() *)
  product()

(*
ocaml src/ocaml/src/sniper.ml
*)

File: src/cpp/src/sniper.cpp
#include <map>
using std::map;

#include "../utils/about_util.cpp"
#include "../utils/tester.cpp"

const string joining = "sniper item-123 1300 0 0 JOINING";
const string bidding = "sniper item-123 1300 1000 1098 BIDDING";
const string winning = "sniper item-123 1300 1098 1098 WINNING";
const string close_msg = "SOLVersion: 1.1; Event: CLOSE;";
const string buy_price_msg = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 1098; Increment: 24; Bidder: other bidder;";
const string price_me = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 1098; Increment: 24; Bidder: sniper;";
const string hi_price = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 1280; Increment: 21; Bidder: some one;";
const string not_e_msg = "SOLVersion: 1.1; Event:; CurrentPrice: 1098; Increment: 24; Bidder: other bidder;";
const string not_i_msg = "SOLVersion: 1.1; Event:PRICE; CurrentPrice: 1098; Bidder: other bidder;";

struct Sniper{
  string name, item, state, msg; int stopPrice, lastPrice, lastBid;
  Sniper (string sniper){
    Ss ss = split(sniper, ' ');
    name = ss.at(0); item = ss.at(1); stopPrice = std::stoi(ss.at(2)); lastPrice = std::stoi(ss.at(3));
    lastBid = std::stoi(ss.at(4)); state = ss.at(5); msg = "toAuctionMessage is None";
  }
  string show(){
    return msg + "\n" + item + " " + to_string(lastPrice) + " " + to_string(lastBid) + " " + state;
  }
  void close(){ state = state == "WINNING" ? "WON" : "LOST";}
  void price(int price, int increment, string bidder){
    if(bidder == name){ lastPrice = price; state = "WINNING"; return;}
    int bid = price + increment; if (bid > stopPrice) { lastPrice = price; state = "LOSING"; return;}
    lastPrice = price; lastBid = bid; state = "BIDDING";
    msg = "SOLVersion: 1.1; Command: BID; Price: " + to_string(bid) + ";";
  }
  void failed(string fail_msg){ lastPrice = 0; lastBid = 0; state = "FAILED"; msg = "failed: " + fail_msg;}
};
void test_s(){ Sniper s = Sniper(joining); cout << s.show() << endl;}

struct Translator{
  string msg, fail_msg, bidder; map<string, string> fields; int price, increment; bool fail = false;
  Translator(const string &message) { msg = message; fail_msg = "[" + msg + "]";}
  void makeFields(){
    Ss fs = split(msg, ';');
    for(int i= 0; i<(int)fs.size(); i++){
      Ss pair = split(fs.at(i), ':');
      string key = trim(pair.at(0)); string val = trim(pair.at(1));
      fields[key] = val;
    }
  }  
  void translate(){ try{ makeFields();} catch (std::out_of_range e) { fail = true;}}
  string get(string key){
    string val = fields[key]; if (val == "") { throw " Missing value for field is [" + key + "]";}
    return val;
  }
  void set_price(){ try {price = stoi(get("CurrentPrice"));} catch(string e){fail = true; fail_msg += e;}}
  void set_increment(){ try {increment = stoi(get("Increment"));} catch(string e){fail = true; fail_msg += e;} }
  void set_bidder(){ try {bidder = get("Bidder");} catch(string e){fail = true; fail_msg += e;}}
  string event(){
    string res;
    try{ res = get("Event");} catch(string e) { fail = true; fail_msg += e;}
    if(res == "CLOSE") { return res;}
    if(res == "PRICE") { set_price(); set_increment(); set_bidder();}
    if (fail) { return "FAILED";}
    return res;
  }
};
void test_t() { Translator t = Translator(price_me); t.translate(); cout << t.event() << endl;}

string domain(string items, string msg){
  Sniper s = Sniper(items); Translator t = Translator(msg); t.translate();
  string event = t.event();
  if(event == "CLOSE") { s.close();}
  if(event == "PRICE") { s.price(t.price, t.increment, t.bidder);}
  if(event == "FAILED") { s.failed(t.fail_msg);}
  return s.show();
}
void test() { string res = domain(bidding, price_me); cout << res << endl;}

string retToSp(const string &str){ Ss res = split(str, '\n'); return trim(res.at(0)) + " " + trim(res.at(1));}

void refactor(){
  string arg = read_file("a:/pj/sniper/rr/req1.txt"), exp = read_file("a:/pj/sniper/rr/res2.txt");
  Ss args = split(arg, '\n'); string act = domain(args.at(0), args.at(1));
  zz_act("Sniper Test", retToSp(act), retToSp(exp), "ここでは失敗メッセージをチェックするよ");
}

void developer(string arg, string exp, string msg){
  Ss args = split(arg, '\n'); string act = domain(args.at(0), args.at(1));
  zz_act("Sniper Test", retToSp(act), retToSp(exp), msg);
}
void develop(){
  string req1 = read_file("a:/pj/sniper/rr/req1.txt"), res1 = read_file("a:/pj/sniper/rr/res1.txt");
  string req2 = read_file("a:/pj/sniper/rr/req2.txt"), res2 = read_file("a:/pj/sniper/rr/res2.txt");
  string req3 = read_file("a:/pj/sniper/rr/req3.txt"), res3 = read_file("a:/pj/sniper/rr/res3.txt");
  string req4 = read_file("a:/pj/sniper/rr/req4.txt"), res4 = read_file("a:/pj/sniper/rr/res4.txt");
  string req5 = read_file("a:/pj/sniper/rr/req5.txt"), res5 = read_file("a:/pj/sniper/rr/res5.txt");
  string req6 = read_file("a:/pj/sniper/rr/req6.txt"), res6 = read_file("a:/pj/sniper/rr/res11.txt");
  developer(req1, res1, "参加要請中に限度額以下の価格情報が来ると入札中になって買い注文をだすよ");
  developer(req2, res2, "参加要請中にオークションが閉まると落札失敗で買い注文は無し");
  developer(req3, res3, "入札中にオークションが閉まると落札失敗で注文はしない");
  developer(req4, res4, "一位入札中にオークションが閉まると落札成功で注文はしない");
  developer(req5, res5, "一位入札中でも限度額以上の価格情報が来ると脱落中で注文はしない");
  developer(req6, res6, "ヘンテコなメッセージが来れば金額は０にしてエラー発生だ");
}

void product(){
  string arg = read_file("toSniper.txt"); Ss args = split(arg, '\n');
  string ans = domain(args.at(0), args.at(1));
  write_file("fromSniper.txt", ans);
}

int main(){
  // test();
  // refactor();
  // develop();
  product();
}

/*
g++ -Wall -Wextra -fexec-charset=cp932 src/cpp/src/sniper.cpp
.\a.exe
*/